server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -750,43 +748,70 @@ protected void doRun() {
                         final org.elasticsearch.script.Metadata originalDocumentMetadata = ingestDocument.getMetadata().clone();
                         // the document listener gives us three-way logic: a document can fail processing (1), or it can
                         // be successfully processed. a successfully processed document can be kept (2) or dropped (3).
-                        final ActionListener<IngestPipelinesExecutionResult> documentListener = ActionListener.runAfter(
-                            new ActionListener<>() {
-                                @Override
-                                public void onResponse(IngestPipelinesExecutionResult result) {
-                                    assert result != null;
```

Comment:
This whole refactor of `documentListener` and `exceptionHandler` isn't really necessary anymore (it's a remnant of one of my previous approaches where it was necessary), but I think it looks a bit cleaner so I decided to leave it in. If people feel differently, I'm happy to revert the changes.